common/src/primitives/per_thousand.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use serialization::{Decode, Encode};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug, Encode, Decode)]
+pub struct PerThousand(#[codec(compact)] u16);
+
+impl PerThousand {
+    pub fn new(value: u16) -> Option<Self> {
+        if value <= 1000 {
```

Comment:
Besides that there are issues with this definition of PerThousand, based on your design this should probably be `value < 1000`.